src/Symfony/Component/HttpClient/Httplug/CorePromise.php

```diff
@@ -0,0 +1,236 @@
+<?php
+
+/*
+ * This file is part of the Symfony package.
+ *
+ * (c) Fabien Potencier <fabien@symfony.com>
+ *
+ * For the full copyright and license information, please view the LICENSE
+ * file that was distributed with this source code.
+ */
+
+namespace Symfony\Component\HttpClient\Httplug;
+
+use Http\Client\Exception;
+use Http\Client\Exception\NetworkException;
+use Http\Client\Exception\RequestException;
+use Http\Promise\Promise;
+use Psr\Http\Message\RequestInterface;
+use Psr\Http\Message\ResponseFactoryInterface;
+use Psr\Http\Message\StreamFactoryInterface;
+use Symfony\Component\HttpClient\Response\ResponseTrait;
+use Symfony\Component\HttpClient\Response\StreamWrapper;
+use Symfony\Contracts\HttpClient\Exception\TransportExceptionInterface;
+use Symfony\Contracts\HttpClient\HttpClientInterface;
+use Symfony\Contracts\HttpClient\ResponseInterface;
+
+/**
+ * @internal
+ *
+ * @author Tobias Nyholm <tobias.nyholm@gmail.com>
+ */
+class CorePromise
+{
+    /**
+     * The Symfony response.
+     *
+     * @var ResponseInterface
+     */
+    private $response;
+
+    /**
+     * @var HttpClientInterface
+     */
+    private $client;
+
+    /**
+     * @var StreamFactoryInterface
+     */
+    private $responseFactory;
+
+    /**
+     * @var StreamFactoryInterface
+     */
+    private $streamFactory;
+
+    /**
+     * The PSR7 request.
+     *
+     * @var RequestInterface
+     */
+    private $psrRequest;
+
+    /**
+     * The HTTPlug response.
+     *
+     * @var ResponseInterface
+     */
+    private $psrResponse;
+
+    /**
+     * Promise state.
+     *
+     * @var string
+     */
+    private $state;
+
+    /**
+     * Exception.
+     *
+     * @var Exception|null
+     */
+    private $exception = null;
+
+    /**
+     * Functions to call when a response will be available.
+     *
+     * @var callable[]
+     */
+    private $onFulfilled = [];
+
+    /**
+     * Functions to call when an error happens.
+     *
+     * @var callable[]
+     */
+    private $onRejected = [];
+
+    public function __construct(
+        RequestInterface $psrRequest,
+        ResponseInterface $symfonyResponse,
+        HttpClientInterface $client,
+        ResponseFactoryInterface $responseFactory,
+        StreamFactoryInterface $streamFactory
+    ) {
+        $this->psrRequest = $psrRequest;
+        $this->response = $symfonyResponse;
+        $this->client = $client;
+        $this->responseFactory = $responseFactory;
+        $this->streamFactory = $streamFactory;
+        $this->state = Promise::PENDING;
+    }
+
+    public function wait()
+    {
+        try {
+            $headers = $this->response->getHeaders(false);
```

Comment:
I'm pretty all other implementations (base on curl) fulfill after the body has been received.
Is this really the semantics of the interface, waiting only for headers?
How do you stream bodies concurrently then?

Reply:
Oh, I just wanted to grab something to get the response. I just definitely get the body here too.

Reply:
actually we could imagine this: when the response is buffered, fulfilling should happen on body completion
when it's not, fulfilling should happen on arrivals of headers so that one can read the streaming body.
There is not way currently to know I think (there may be one but I don't know it yet :) )